src/tests/useQuery.test.js

```diff
@@ -249,6 +249,37 @@ describe('useQuery', () => {
     expect(queryFn).toHaveBeenCalledTimes(2)
   })
 
+  it('should not retry if retryChecker returns `false`', async () => {
+    const queryFn = jest.fn()
+
+    queryFn.mockImplementation(() => {
+      return Promise.reject('NoRetry')
+    })
+
+    function Page() {
+      const { status, failureCount } = useQuery('test', queryFn, {
+        retryDelay: 1,
+        retry: (failureCount, error) => error !== 'NoRetry',
```

Comment:
This test will pass even without this feature implemented:
`retry` is a function, therefore `retry === true` is `false` and `query.state.failureCount <= query.config.retry` is false as well, so query won't be refetched
https://github.com/tannerlinsley/react-query/blob/2b6a711cf38acaa891bf0db3a6d871247abe9b9a/src/queryCache.js#L315-L319

My suggestion is to allow retry once and expect query to fail 2 times